In [1]:
import pandas as pd

orders = pd.read_csv("data/olist_orders_dataset.csv")
customers = pd.read_csv("data/olist_customers_dataset.csv")
order_items = pd.read_csv("data/olist_order_items_dataset.csv")
products = pd.read_csv("data/olist_products_dataset.csv")

##  Dataset Overview

<Dataset Overview>

In [2]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [3]:
orders.shape

(99441, 8)

In [4]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [5]:
orders.dtypes

order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object

In [6]:
orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [7]:
orders.duplicated().sum()

np.int64(0)

##  Data Cleaning

In [8]:
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])
orders["order_approved_at"]=pd.to_datetime(orders["order_approved_at"])
orders["order_delivered_carrier_date"]=pd.to_datetime(orders["order_delivered_carrier_date"])
orders["order_delivered_customer_date"]=pd.to_datetime(orders["order_delivered_customer_date"])
orders["order_estimated_delivery_date"]=pd.to_datetime(orders["order_estimated_delivery_date"])

In [9]:
orders.dtypes

order_id                                    str
customer_id                                 str
order_status                                str
order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [10]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [11]:
for name, df in {
    "customers": customers,
    "order_items": order_items,
    "products": products
}.items():

    print(f"===== {name.upper()} OVERVIEW =====")

    print("\nHEAD:")
    print(df.head())

    print("\nSHAPE:")
    print(df.shape)

    print("\nCOLUMNS:")
    print(df.columns)

    print("\nDTYPES:")
    print(df.dtypes)

    print("\nMISSING VALUES:")
    print(df.isna().sum())

    print("\nDUPLICATES:")
    print(df.duplicated().sum())

    print("\n")

===== CUSTOMERS OVERVIEW =====

HEAD:
                        customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  

SHAPE:
(99441, 5)

COLUMNS:
Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'custom

In [12]:
orders_customers = pd.merge(orders, customers, on="customer_id", how="left")

In [13]:
print("orders:", orders.shape)
print("orders_customers:", orders_customers.shape)

orders: (99441, 8)
orders_customers: (99441, 12)


In [14]:
orders_customers.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,sao paulo,SP
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,barreiras,BA
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,vianopolis,GO
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,7c142cf63193a1473d2e66489a9ae977,59296,sao goncalo do amarante,RN
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,72632f0f9dd73dfee390c9b22eb56dd6,9195,santo andre,SP


## Data Preparation and Merging

In [15]:
sales_data = pd.merge(
    orders_customers,
    order_items,
    on="order_id",
    how="left"
)

In [16]:
print("orders_customers:", orders_customers.shape)
print("sales_data:", sales_data.shape)

orders_customers: (99441, 12)
sales_data: (113425, 18)


In [17]:
final_data = pd.merge(
    sales_data,
    products,
    on="product_id",
    how="left"
)

In [18]:
print("sales_data:", sales_data.shape)
print("final_data:", final_data.shape)

sales_data: (113425, 18)
final_data: (113425, 26)


In [19]:
final_data.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,customer_unique_id,customer_zip_code_prefix,...,price,freight_value,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,7c396fd4830fd04220f754e42b4e5bff,3149,...,29.99,8.72,utilidades_domesticas,40.0,268.0,4.0,500.0,19.0,8.0,13.0
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,af07308b275d755c9edb36a90c618231,47813,...,118.70,22.76,perfumaria,29.0,178.0,1.0,400.0,19.0,13.0,19.0
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,3a653a41f6f9fc3d2a113cf8398680e8,75265,...,159.90,19.22,automotivo,46.0,232.0,1.0,420.0,24.0,19.0,21.0
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,7c142cf63193a1473d2e66489a9ae977,59296,...,45.00,27.20,pet_shop,59.0,468.0,3.0,450.0,30.0,10.0,20.0
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,72632f0f9dd73dfee390c9b22eb56dd6,9195,...,19.90,8.72,papelaria,38.0,316.0,4.0,250.0,51.0,15.0,15.0


## Business analysis

### 5.1 Total Revenue

In [20]:
total_revenue = final_data["price"].sum()
total_revenue

np.float64(13591643.7)

### 5.2 Revenue by State

In [21]:
top_states = (
    final_data
    .groupby("customer_state")["price"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_states

customer_state
SP    5202955.05
RJ    1824092.67
MG    1585308.03
RS     750304.02
PR     683083.76
SC     520553.34
BA     511349.99
DF     302603.94
GO     294591.95
ES     275037.31
Name: price, dtype: float64

In [35]:
top_states_pct = (top_states / final_data["price"].sum()) * 100

print(top_states_pct.round(2))

customer_state
SP    38.28
RJ    13.42
MG    11.66
RS     5.52
PR     5.03
SC     3.83
BA     3.76
DF     2.23
GO     2.17
ES     2.02
Name: price, dtype: float64


### 5.3 Average Order Value

In [22]:
avg_order_value = (
    final_data
    .groupby("order_id")["price"]
    .sum()
    .mean()
)

avg_order_value

np.float64(136.68048088816482)

### 5.4 Order Status Distribution

In [23]:
order_status_distribution = orders["order_status"].value_counts()
order_status_distribution

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [33]:

order_status_pct = orders["order_status"].value_counts(normalize=True) * 100

print(order_status_pct.round(2))

order_status
delivered      97.02
shipped         1.11
canceled        0.63
unavailable     0.61
invoiced        0.32
processing      0.30
created         0.01
approved        0.00
Name: proportion, dtype: float64


### 5.5 Top Customers

In [24]:
top_customers = (
    final_data
    .groupby("customer_unique_id")["price"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_customers

customer_unique_id
0a0a92112bd4c708ca5fde585afaa872    13440.0
da122df9eeddfedc1dc1f5349a1a690c     7388.0
763c8b1c9c68a0229c42c9fc6f662b93     7160.0
dc4802a71eae9be1dd28f5d788ceb526     6735.0
459bef486812aa25204be022145caa62     6729.0
ff4159b92c40ebe40454e3e6a7c35ed6     6499.0
4007669dec559734d6f53e029e360987     5934.6
eebb5dda148d3893cdaf5b5ca3040ccb     4690.0
5d0a2980b292d049061542014e8960bf     4599.9
48e1ac109decbb87765a3eade6854098     4590.0
Name: price, dtype: float64

### 5.6 Monthly Revenue

In [25]:
final_data["month"] = final_data["order_purchase_timestamp"].dt.to_period("M")

In [26]:
final_data[["order_purchase_timestamp", "month"]].head()

,order_purchase_timestamp,month
0,2017-10-02 10:56:33,2017-10
1,2018-07-24 20:41:37,2018-07
2,2018-08-08 08:38:49,2018-08
3,2017-11-18 19:28:06,2017-11
4,2018-02-13 21:18:39,2018-02


In [27]:
monthly_revenue = (
    final_data
    .groupby("month")["price"]
    .sum()
    .sort_index()
)

monthly_revenue

month
2016-09        267.36
2016-10      49507.66
2016-12         10.90
2017-01     120312.87
2017-02     247303.02
2017-03     374344.30
2017-04     359927.23
2017-05     506071.14
2017-06     433038.60
2017-07     498031.48
2017-08     573971.68
2017-09     624401.69
2017-10     664219.43
2017-11    1010271.37
2017-12     743914.17
2018-01     950030.36
2018-02     844178.71
2018-03     983213.44
2018-04     996647.75
2018-05     996517.68
2018-06     865124.31
2018-07     895507.22
2018-08     854686.33
2018-09        145.00
2018-10          0.00
Freq: M, Name: price, dtype: float64

### 5.7 Top Product Categories

In [28]:
top_categories = (
    final_data
    .groupby("product_category_name")["price"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_categories

product_category_name
beleza_saude              1258681.34
relogios_presentes        1205005.68
cama_mesa_banho           1036988.68
esporte_lazer              988048.97
informatica_acessorios     911954.32
moveis_decoracao           729762.49
cool_stuff                 635290.85
utilidades_domesticas      632248.66
automotivo                 592720.11
ferramentas_jardim         485256.46
Name: price, dtype: float64

### 5.8 Category Performance

In [29]:
category_performance = (
    final_data
    .groupby("product_category_name")
    .agg(
        total_revenue=("price", "sum"),
        item_count=("order_item_id", "count")
    )
    .sort_values("total_revenue", ascending=False)
    .head(10)
)

category_performance

,total_revenue,item_count
product_category_name,,
beleza_saude,1258681.34,9670
relogios_presentes,1205005.68,5991
cama_mesa_banho,1036988.68,11115
esporte_lazer,988048.97,8641
informatica_acessorios,911954.32,7827
moveis_decoracao,729762.49,8334
cool_stuff,635290.85,3796
utilidades_domesticas,632248.66,6964
automotivo,592720.11,4235


##  Key Insights

- The majority of orders were successfully delivered (97.02%).

- Revenue is concentrated in several states. SP has the highest share of total revenue (38.28%), followed by RJ (13.42%) and MG (11.66%). Together, these three states account for more than half of total revenue.

- The average order value is around $136.68.

- There is a noticeable difference in spending between customers, with some customers spending significantly more than others.

- Some product categories generate significantly more revenue than others. The top categories are `beleza_saude` (1,258,681.34), `relogios_presentes` (1,205,005.68), and `cama_mesa_banho` (1,036,988.68).

- Monthly revenue changes noticeably over time. The peak was in November 2017, while October 2016 had the lowest meaningful revenue, excluding October 2018.

## CONCLUSION

The analysis showed how customer location, product category and time are related to revenue.

The project also provided practical experience with cleaning data, merging multiple tables and using Pandas for business analysis.